From 6f9e2b1500d53bdcfacbe8b8428e93ec5a033178 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:04:13 -0400
Subject: [usb] xhci: fix incorrect type in assignment in xhci_count_num_dropped_endpoints()

Message-id: <1410887463-20674-182-git-send-email-dzickus@redhat.com>
Patchwork-id: 95250
O-Subject: [RHEL7 PATCH 181/591] xhci: fix incorrect type in assignment in xhci_count_num_dropped_endpoints()
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 78d1ff025698e1a70e3fef38342648d4dc826cf6
Author: Xenia Ragiadakou <burzalodowa@gmail.com>
Date:   Mon Sep 9 21:03:07 2013 +0300

    xhci: fix incorrect type in assignment in xhci_count_num_dropped_endpoints()

    The fields 'add_flags' and 'drop_flags' in struct xhci_input_control_ctx
    have type __le32 and need to be converted to CPU byteorder before being
    used to derive the number of dropped endpoints.
    This bug was found using sparse.

    This patch is not suitable for stable, since the bug would only be
    triggered on big endian systems, and the code only runs for Intel xHCI
    host controllers, which are always integrated into little endian
    systems.

    Signed-off-by: Xenia Ragiadakou <burzalodowa@gmail.com>
    Signed-off-by: Sarah Sharp <sarah.a.sharp@linux.intel.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/host/xhci.c b/drivers/usb/host/xhci.c
index b7289e9..900ba36 100644
--- a/drivers/usb/host/xhci.c
+++ b/drivers/usb/host/xhci.c
@@ -1909,8 +1909,8 @@ static unsigned int xhci_count_num_dropped_endpoints(struct xhci_hcd *xhci,
  u32 valid_add_flags;
  u32 valid_drop_flags;
 
- valid_add_flags = ctrl_ctx->add_flags >> 2;
- valid_drop_flags = ctrl_ctx->drop_flags >> 2;
+ valid_add_flags = le32_to_cpu(ctrl_ctx->add_flags) >> 2;
+ valid_drop_flags = le32_to_cpu(ctrl_ctx->drop_flags) >> 2;
 
  return hweight32(valid_drop_flags) -
   hweight32(valid_add_flags & valid_drop_flags);
-- 
1.7.1